# Data Quality Assessment

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Analysing first sheet

In [2]:
# this sheet contains multiple sheets so pass the sheet name argument to read excel 
df = pd.read_excel(r"../KPMG_VI_New_raw_data_update_final.xlsx",sheet_name='Transactions')
df.head()

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                     transaction_id                                                                                                                              
1                                                  1                                                                                                                              
2                                                  2                                                                                                                              
3                                                  3                                                                                                                              
4                                                  4                                                                                                                              

   Unnamed: 1   Unnamed: 2           Unnamed: 3    Unnamed: 4    Unnamed: 5  \
0  product_id  customer_id     transaction_date  online_order  order_status   
1           2         2950  2017-02-25 00:00:00         False      Approved   
2           3         3120  2017-05-21 00:00:00          True      Approved   
3          37          402  2017-10-16 00:00:00         False      Approved   
4          88         3135  2017-08-31 00:00:00         False      Approved   

       Unnamed: 6    Unnamed: 7     Unnamed: 8    Unnamed: 9  ... Unnamed: 16  \
0           brand  product_line  product_class  product_size  ...         NaN   
1           Solex      Standard         medium        medium  ...         NaN   
2   Trek Bicycles      Standard         medium         large  ...         NaN   
3      OHM Cycles      Standard            low        medium  ...         NaN   
4  Norco Bicycles      Standard         medium        medium  ...         NaN   

  Unnamed: 17 Unnamed: 18  Unnamed: 19  Unnamed: 20  Unnamed: 21  Unnamed: 22  \
0         NaN         NaN          NaN          NaN          NaN          NaN   
1         NaN         NaN          NaN          NaN          NaN          NaN   
2         NaN         NaN          NaN          NaN          NaN          NaN   
3         NaN         NaN          NaN          NaN          NaN          NaN   
4         NaN         NaN          NaN          NaN          NaN          NaN   

   Unnamed: 23  Unnamed: 24  Unnamed: 25  
0          NaN          NaN          NaN  
1          NaN          NaN          NaN  
2          NaN          NaN          NaN  
3          NaN          NaN          NaN  
4          NaN          NaN          NaN  

[5 rows x 26 columns]

###  Cleaning the header row ,index column etc

In [3]:
# first should be column names for the dataset
col = df.loc[0]
df = df[1:]
df.rename(columns=col,inplace=True)
df.head()

transaction_id product_id customer_id     transaction_date online_order  \
1              1          2        2950  2017-02-25 00:00:00        False   
2              2          3        3120  2017-05-21 00:00:00         True   
3              3         37         402  2017-10-16 00:00:00        False   
4              4         88        3135  2017-08-31 00:00:00        False   
5              5         78         787  2017-10-01 00:00:00         True   

  order_status           brand product_line product_class product_size  ...  \
1     Approved           Solex     Standard        medium       medium  ...   
2     Approved   Trek Bicycles     Standard        medium        large  ...   
3     Approved      OHM Cycles     Standard           low       medium  ...   
4     Approved  Norco Bicycles     Standard        medium       medium  ...   
5     Approved  Giant Bicycles     Standard        medium        large  ...   

   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 26 columns]

In [4]:
# out of 26 columns only 13 columns were useful so remove 
df = df.iloc[:,:13]
df.head()

transaction_id product_id customer_id     transaction_date online_order  \
1              1          2        2950  2017-02-25 00:00:00        False   
2              2          3        3120  2017-05-21 00:00:00         True   
3              3         37         402  2017-10-16 00:00:00        False   
4              4         88        3135  2017-08-31 00:00:00        False   
5              5         78         787  2017-10-01 00:00:00         True   

  order_status           brand product_line product_class product_size  \
1     Approved           Solex     Standard        medium       medium   
2     Approved   Trek Bicycles     Standard        medium        large   
3     Approved      OHM Cycles     Standard           low       medium   
4     Approved  Norco Bicycles     Standard        medium       medium   
5     Approved  Giant Bicycles     Standard        medium        large   

  list_price standard_cost product_first_sold_date  
1      71.49         53.62                   41245  
2    2091.47        388.92                   41701  
3    1793.43        248.82                   36361  
4    1198.46         381.1                   36145  
5     1765.3        709.48                   42226

In [5]:
# setting transaction id as index
df.set_index('transaction_id',inplace = True)
df.head()

product_id customer_id     transaction_date online_order  \
transaction_id                                                            
1                       2        2950  2017-02-25 00:00:00        False   
2                       3        3120  2017-05-21 00:00:00         True   
3                      37         402  2017-10-16 00:00:00        False   
4                      88        3135  2017-08-31 00:00:00        False   
5                      78         787  2017-10-01 00:00:00         True   

               order_status           brand product_line product_class  \
transaction_id                                                           
1                  Approved           Solex     Standard        medium   
2                  Approved   Trek Bicycles     Standard        medium   
3                  Approved      OHM Cycles     Standard           low   
4                  Approved  Norco Bicycles     Standard        medium   
5                  Approved  Giant Bicycles     Standard        medium   

               product_size list_price standard_cost product_first_sold_date  
transaction_id                                                                
1                    medium      71.49         53.62                   41245  
2                     large    2091.47        388.92                   41701  
3                    medium    1793.43        248.82                   36361  
4                    medium    1198.46         381.1                   36145  
5                     large     1765.3        709.48                   42226

### Applying transformation of columns

In [6]:
df.dtypes

product_id                 object
customer_id                object
transaction_date           object
online_order               object
order_status               object
brand                      object
product_line               object
product_class              object
product_size               object
list_price                 object
standard_cost              object
product_first_sold_date    object
dtype: object

#### converting date time columns 
making changes in timestamps and other date columns

In [7]:
# we see transaction_date column as object dtype, we will change it to datetime dtype
df.transaction_date = pd.to_datetime(df.transaction_date)

#product first_sold_date is a timestamp in object dtype
df.product_first_sold_date = pd.to_datetime(df.product_first_sold_date).dt.date

In [8]:
df = df.astype({'product_id':int,"customer_id":int,'list_price':float,"standard_cost":float,'online_order':bool})

In [9]:
df.dtypes

product_id                          int32
customer_id                         int32
transaction_date           datetime64[ns]
online_order                         bool
order_status                       object
brand                              object
product_line                       object
product_class                      object
product_size                       object
list_price                        float64
standard_cost                     float64
product_first_sold_date            object
dtype: object

In [10]:
df.head()

product_id  customer_id transaction_date  online_order  \
transaction_id                                                           
1                        2         2950       2017-02-25         False   
2                        3         3120       2017-05-21          True   
3                       37          402       2017-10-16         False   
4                       88         3135       2017-08-31         False   
5                       78          787       2017-10-01          True   

               order_status           brand product_line product_class  \
transaction_id                                                           
1                  Approved           Solex     Standard        medium   
2                  Approved   Trek Bicycles     Standard        medium   
3                  Approved      OHM Cycles     Standard           low   
4                  Approved  Norco Bicycles     Standard        medium   
5                  Approved  Giant Bicycles     Standard        medium   

               product_size  list_price  standard_cost product_first_sold_date  
transaction_id                                                                  
1                    medium       71.49          53.62              1970-01-01  
2                     large     2091.47         388.92              1970-01-01  
3                    medium     1793.43         248.82              1970-01-01  
4                    medium     1198.46         381.10              1970-01-01  
5                     large     1765.30         709.48              1970-01-01

In [11]:
df.describe()

product_id   customer_id    list_price  standard_cost
count  20000.00000  20000.000000  20000.000000   19803.000000
mean      45.36465   1738.246050   1107.829449     556.046951
std       30.75359   1011.951046    582.825242     405.955660
min        0.00000      1.000000     12.010000       7.210000
25%       18.00000    857.750000    575.270000     215.140000
50%       44.00000   1736.000000   1163.890000     507.580000
75%       72.00000   2613.000000   1635.300000     795.100000
max      100.00000   5034.000000   2091.470000    1759.850000

#### Dealing with null values

In [12]:
df.isnull().sum()

product_id                   0
customer_id                  0
transaction_date             0
online_order                 0
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [13]:
df.online_order = df.online_order.map({True:1,False:0})
df.online_order = df.online_order.astype(int)

In [14]:
df.dtypes

product_id                          int32
customer_id                         int32
transaction_date           datetime64[ns]
online_order                        int32
order_status                       object
brand                              object
product_line                       object
product_class                      object
product_size                       object
list_price                        float64
standard_cost                     float64
product_first_sold_date            object
dtype: object

In [15]:
df.order_status.value_counts()

Approved     19821
Cancelled      179
Name: order_status, dtype: int64

In [16]:
df[['brand','order_status']].value_counts(normalize=True)

brand           order_status
Solex           Approved        0.212645
Giant Bicycles  Approved        0.165783
WeareA2B        Approved        0.164874
OHM Cycles      Approved        0.152300
Trek Bicycles   Approved        0.149725
Norco Bicycles  Approved        0.145685
Solex           Cancelled       0.002121
WeareA2B        Cancelled       0.001515
Giant Bicycles  Cancelled       0.001464
OHM Cycles      Cancelled       0.001363
Norco Bicycles  Cancelled       0.001262
Trek Bicycles   Cancelled       0.001262
dtype: float64

In [17]:
df[df.index.duplicated()]

Empty DataFrame
Columns: [product_id, customer_id, transaction_date, online_order, order_status, brand, product_line, product_class, product_size, list_price, standard_cost, product_first_sold_date]
Index: []

In [18]:
df.isnull().sum()

product_id                   0
customer_id                  0
transaction_date             0
online_order                 0
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

We can either remove all the 197 rows or fill them most repeated value of that column

In [19]:
df.product_line.mode()

0    Standard
dtype: object

In [20]:
df.product_class.value_counts()

medium    13826
high       3013
low        2964
Name: product_class, dtype: int64

In [21]:
df.product_size.value_counts()

medium    12990
large      3976
small      2837
Name: product_size, dtype: int64

In [22]:
df.standard_cost.value_counts()

388.920000    465
954.820000    396
53.620000     274
161.600000    235
260.140000    233
             ... 
151.960000    124
206.350000    114
667.400024      1
270.299988      1
312.735016      1
Name: standard_cost, Length: 103, dtype: int64

In [23]:
np.mean(df.standard_cost)

556.0469512209312

In [24]:
df.isna().sum()

product_id                   0
customer_id                  0
transaction_date             0
online_order                 0
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [25]:
df.fillna({'brand':df.brand.mode()[0],"product_class":df.product_class.mode()[0],"product_line":df.product_line.mode()[0],"product_size":df.product_size.mode()[0],"standard_cost":np.mean(df.standard_cost),"product_first_sold_date":pd.to_datetime(df.product_first_sold_date).dt.year.mode()[0]},inplace = True)

In [26]:
# pd.to_datetime(df.product_first_sold_date).dt.year.mode()[0]

In [27]:
df.brand.mode()

0    Solex
dtype: object

In [28]:
df[df.index.duplicated()]

Empty DataFrame
Columns: [product_id, customer_id, transaction_date, online_order, order_status, brand, product_line, product_class, product_size, list_price, standard_cost, product_first_sold_date]
Index: []

No duplicated columns are present in the the dataset so no need to delete any rows

In [29]:
df.isna().sum()

product_id                 0
customer_id                0
transaction_date           0
online_order               0
order_status               0
brand                      0
product_line               0
product_class              0
product_size               0
list_price                 0
standard_cost              0
product_first_sold_date    0
dtype: int64

## Analyzing the second sheet 

In [30]:
# this sheet contains multiple sheets so pass the sheet name argument to read excel 
df1 = pd.read_excel(r"../KPMG_VI_New_raw_data_update_final.xlsx",sheet_name='NewCustomerList')
df1.head()

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                         first_name                                                                                                                              
1                                            Chickie                                                                                                                              
2                                              Morly                                                                                                                              
3                                            Ardelis                                                                                                                              
4                                             Lucine                                                                                                                              

  Unnamed: 1 Unnamed: 2                           Unnamed: 3  \
0  last_name     gender  past_3_years_bike_related_purchases   
1    Brister       Male                                   86   
2     Genery       Male                                   69   
3  Forrester     Female                                   10   
4      Stutt     Female                                   64   

            Unnamed: 4                  Unnamed: 5             Unnamed: 6  \
0                  DOB                   job_title  job_industry_category   
1           1957-07-12             General Manager          Manufacturing   
2           1970-03-22         Structural Engineer               Property   
3  1974-08-28 00:00:00      Senior Cost Accountant     Financial Services   
4           1979-01-28  Account Representative III          Manufacturing   

          Unnamed: 7          Unnamed: 8 Unnamed: 9  ... Unnamed: 13  \
0     wealth_segment  deceased_indicator   owns_car  ...       state   
1      Mass Customer                   N        Yes  ...         QLD   
2      Mass Customer                   N         No  ...         NSW   
3  Affluent Customer                   N         No  ...         VIC   
4  Affluent Customer                   N        Yes  ...         QLD   

  Unnamed: 14         Unnamed: 15 Unnamed: 16 Unnamed: 17 Unnamed: 18  \
0     country  property_valuation         NaN         NaN         NaN   
1   Australia                   6        0.56      0.7000      0.8750   
2   Australia                  11        0.89      0.8900      1.1125   
3   Australia                   5        1.01      1.0100      1.0100   
4   Australia                   1        0.87      1.0875      1.0875   

   Unnamed: 19  Unnamed: 20  Unnamed: 21  Unnamed: 22  
0          NaN          NaN         Rank        Value  
1     0.743750          1.0            1      1.71875  
2     0.945625          1.0            1      1.71875  
3     1.010000          1.0            1      1.71875  
4     1.087500          4.0            4     1.703125  

[5 rows x 23 columns]

In [31]:
col_name = df1.iloc[0]
df1=df1.loc[1:]
df1.rename(columns = col_name,inplace = True)

In [32]:
df1.shape

(1000, 23)

In [33]:
df1.columns

Index([                         'first_name',
                                 'last_name',
                                    'gender',
       'past_3_years_bike_related_purchases',
                                       'DOB',
                                 'job_title',
                     'job_industry_category',
                            'wealth_segment',
                        'deceased_indicator',
                                  'owns_car',
                                    'tenure',
                                   'address',
                                  'postcode',
                                     'state',
                                   'country',
                        'property_valuation',
                                         nan,
                                         nan,
                                         nan,
                                         nan,
                                         nan,
                                  

In [34]:
pd.set_option("display.max_columns",23)

In [35]:
df1.head()

first_name  last_name  gender past_3_years_bike_related_purchases  \
1    Chickie    Brister    Male                                  86   
2      Morly     Genery    Male                                  69   
3    Ardelis  Forrester  Female                                  10   
4     Lucine      Stutt  Female                                  64   
5    Melinda     Hadlee  Female                                  34   

                   DOB                   job_title job_industry_category  \
1           1957-07-12             General Manager         Manufacturing   
2           1970-03-22         Structural Engineer              Property   
3  1974-08-28 00:00:00      Senior Cost Accountant    Financial Services   
4           1979-01-28  Account Representative III         Manufacturing   
5           1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car tenure              address  \
1      Mass Customer                  N      Yes     14     45 Shopko Center   
2      Mass Customer                  N       No     16    14 Mccormick Park   
3  Affluent Customer                  N       No     10  5 Colorado Crossing   
4  Affluent Customer                  N      Yes      5   207 Annamark Plaza   
5  Affluent Customer                  N       No     19    115 Montana Place   

  postcode state    country property_valuation   NaN     NaN     NaN  \
1     4500   QLD  Australia                  6  0.56  0.7000  0.8750   
2     2113   NSW  Australia                 11  0.89  0.8900  1.1125   
3     3505   VIC  Australia                  5  1.01  1.0100  1.0100   
4     4814   QLD  Australia                  1  0.87  1.0875  1.0875   
5     2093   NSW  Australia                  9  0.52  0.5200  0.6500   

        NaN  NaN Rank     Value  
1  0.743750  1.0    1   1.71875  
2  0.945625  1.0    1   1.71875  
3  1.010000  1.0    1   1.71875  
4  1.087500  4.0    4  1.703125  
5  0.650000  4.0    4  1.703125

In [36]:
#there are some nan columns in the dataset 
df1 = df1.loc[:,df1.columns.notnull()]

In [37]:
df1.shape

(1000, 18)

In [38]:
df1.columns[16:21]

Index(['Rank', 'Value'], dtype='object')

In [39]:
df1

first_name  last_name  gender past_3_years_bike_related_purchases  \
1       Chickie    Brister    Male                                  86   
2         Morly     Genery    Male                                  69   
3       Ardelis  Forrester  Female                                  10   
4        Lucine      Stutt  Female                                  64   
5       Melinda     Hadlee  Female                                  34   
...         ...        ...     ...                                 ...   
996   Ferdinand  Romanetti    Male                                  60   
997        Burk    Wortley    Male                                  22   
998    Melloney      Temby  Female                                  17   
999      Dickie    Cubbini    Male                                  30   
1000      Sylas    Duffill    Male                                  56   

                      DOB                     job_title job_industry_category  \
1              1957-07-12               General Manager         Manufacturing   
2              1970-03-22           Structural Engineer              Property   
3     1974-08-28 00:00:00        Senior Cost Accountant    Financial Services   
4              1979-01-28    Account Representative III         Manufacturing   
5              1965-09-21             Financial Analyst    Financial Services   
...                   ...                           ...                   ...   
996            1959-10-07                     Paralegal    Financial Services   
997            2001-10-17        Senior Sales Associate                Health   
998            1954-10-05  Budget/Accounting Analyst IV    Financial Services   
999            1952-12-17             Financial Advisor    Financial Services   
1000           1955-10-02           Staff Accountant IV              Property   

         wealth_segment deceased_indicator owns_car tenure  \
1         Mass Customer                  N      Yes     14   
2         Mass Customer                  N       No     16   
3     Affluent Customer                  N       No     10   
4     Affluent Customer                  N      Yes      5   
5     Affluent Customer                  N       No     19   
...                 ...                ...      ...    ...   
996   Affluent Customer                  N       No      9   
997       Mass Customer                  N       No      6   
998   Affluent Customer                  N      Yes     15   
999       Mass Customer                  N      Yes     19   
1000      Mass Customer                  N      Yes     14   

                       address postcode state    country property_valuation  \
1             45 Shopko Center     4500   QLD  Australia                  6   
2            14 Mccormick Park     2113   NSW  Australia                 11   
3          5 Colorado Crossing     3505   VIC  Australia                  5   
4           207 Annamark Plaza     4814   QLD  Australia                  1   
5            115 Montana Place     2093   NSW  Australia                  9   
...                        ...      ...   ...        ...                ...   
996                2 Sloan Way     2200   NSW  Australia                  7   
997          04 Union Crossing     2196   NSW  Australia                 10   
998   33475 Fair Oaks Junction     4702   QLD  Australia                  2   
999         57666 Victoria Way     4215   QLD  Australia                  2   
1000        21875 Grover Drive     2010   NSW  Australia                  9   

      Rank     Value  
1        1   1.71875  
2        1   1.71875  
3        1   1.71875  
4        4  1.703125  
5        4  1.703125  
...    ...       ...  
996    996     0.374  
997    997     0.357  
998    997     0.357  
999    997     0.357  
1000  1000      0.34  

[1000 rows x 18 columns]

In [40]:
df1.job_industry_category.unique()

array(['Manufacturing', 'Property', 'Financial Services', 'Entertainment',
       'Retail', 'IT', 'Telecommunications', 'Health', nan, 'Argiculture'],
      dtype=object)

In [41]:
df1.dtypes

first_name                             object
last_name                              object
gender                                 object
past_3_years_bike_related_purchases    object
DOB                                    object
job_title                              object
job_industry_category                  object
wealth_segment                         object
deceased_indicator                     object
owns_car                               object
tenure                                 object
address                                object
postcode                               object
state                                  object
country                                object
property_valuation                     object
Rank                                   object
Value                                  object
dtype: object

In [42]:
df1.rename({"past_3_years_bike_related_purchases":'other_purchases',"job_industry_category":"job_industry"},axis = 1,inplace=True)

In [43]:
df1 = df1.astype({"other_purchases":int,"tenure":int,"property_valuation" :int,"Rank":int,"Value":float})

In [44]:
df1.Value = df1.Value.round(3)

In [45]:
df1.head()

first_name  last_name  gender  other_purchases                  DOB  \
1    Chickie    Brister    Male               86           1957-07-12   
2      Morly     Genery    Male               69           1970-03-22   
3    Ardelis  Forrester  Female               10  1974-08-28 00:00:00   
4     Lucine      Stutt  Female               64           1979-01-28   
5    Melinda     Hadlee  Female               34           1965-09-21   

                    job_title        job_industry     wealth_segment  \
1             General Manager       Manufacturing      Mass Customer   
2         Structural Engineer            Property      Mass Customer   
3      Senior Cost Accountant  Financial Services  Affluent Customer   
4  Account Representative III       Manufacturing  Affluent Customer   
5           Financial Analyst  Financial Services  Affluent Customer   

  deceased_indicator owns_car  tenure              address postcode state  \
1                  N      Yes      14     45 Shopko Center     4500   QLD   
2                  N       No      16    14 Mccormick Park     2113   NSW   
3                  N       No      10  5 Colorado Crossing     3505   VIC   
4                  N      Yes       5   207 Annamark Plaza     4814   QLD   
5                  N       No      19    115 Montana Place     2093   NSW   

     country  property_valuation  Rank  Value  
1  Australia                   6     1  1.719  
2  Australia                  11     1  1.719  
3  Australia                   5     1  1.719  
4  Australia                   1     4  1.703  
5  Australia                   9     4  1.703

In [46]:
df1.isnull().sum()

first_name              0
last_name              29
gender                  0
other_purchases         0
DOB                    17
job_title             106
job_industry          165
wealth_segment          0
deceased_indicator      0
owns_car                0
tenure                  0
address                 0
postcode                0
state                   0
country                 0
property_valuation      0
Rank                    0
Value                   0
dtype: int64

In [47]:
#if person dont have last name fill it with "." or we can fill it with first name
df1.last_name = df1.last_name.fillna(".")

In [48]:
df1.job_title.value_counts()

Associate Professor          15
Software Consultant          14
Environmental Tech           14
Chief Design Engineer        13
VP Sales                     12
                             ..
Web Developer IV              1
Web Developer I               1
Safety Technician IV          1
Database Administrator IV     1
Accounting Assistant IV       1
Name: job_title, Length: 184, dtype: int64

In [49]:
df1.DOB = pd.to_datetime(df1.DOB)

In [50]:
pd.to_datetime(df1.DOB).dt.year.mode()

0    1974.0
dtype: float64

In [51]:
df1.DOB.fillna('1974-01-01',inplace = True)

In [52]:
df1["year"] = df1.DOB.apply(lambda x : str(x).split("-")[0])

In [53]:
df1.year.sort_values()

586    1938
407    1938
886    1938
798    1938
266    1938
       ... 
142    2001
476    2001
397    2002
355    2002
764    2002
Name: year, Length: 1000, dtype: object

In [54]:
df1.dropna(inplace=True)

In [55]:
df1.isna().sum()

first_name            0
last_name             0
gender                0
other_purchases       0
DOB                   0
job_title             0
job_industry          0
wealth_segment        0
deceased_indicator    0
owns_car              0
tenure                0
address               0
postcode              0
state                 0
country               0
property_valuation    0
Rank                  0
Value                 0
year                  0
dtype: int64

## Analyzing 3rd sheet

In [56]:
df2 = pd.read_excel("../KPMG_VI_New_raw_data_update_final.xlsx",sheet_name='CustomerDemographic')

In [57]:
df2.head()

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                        customer_id                                                                                                                              
1                                                  1                                                                                                                              
2                                                  2                                                                                                                              
3                                                  3                                                                                                                              
4                                                  4                                                                                                                              

   Unnamed: 1 Unnamed: 2 Unnamed: 3                           Unnamed: 4  \
0  first_name  last_name     gender  past_3_years_bike_related_purchases   
1     Laraine  Medendorp          F                                   93   
2         Eli    Bockman       Male                                   81   
3       Arlin     Dearle       Male                                   61   
4      Talbot        NaN       Male                                   33   

            Unnamed: 5              Unnamed: 6             Unnamed: 7  \
0                  DOB               job_title  job_industry_category   
1  1953-10-12 00:00:00     Executive Secretary                 Health   
2  1980-12-16 00:00:00  Administrative Officer     Financial Services   
3  1954-01-20 00:00:00      Recruiting Manager               Property   
4  1961-10-03 00:00:00                     NaN                     IT   

       Unnamed: 8          Unnamed: 9  \
0  wealth_segment  deceased_indicator   
1   Mass Customer                   N   
2   Mass Customer                   N   
3   Mass Customer                   N   
4   Mass Customer                   N   

                                         Unnamed: 10  ... Unnamed: 15  \
0                                            default  ...         NaN   
1                                                 "'  ...         NaN   
2                       <script>alert('hi')</script>  ...         NaN   
3                                2018-02-01 00:00:00  ...         NaN   
4  () { _; } >_[$($())] { touch /tmp/blns.shellsh...  ...         NaN   

  Unnamed: 16  Unnamed: 17  Unnamed: 18  Unnamed: 19  Unnamed: 20  \
0         NaN          NaN          NaN          NaN          NaN   
1         NaN          NaN          NaN          NaN          NaN   
2         NaN          NaN          NaN          NaN          NaN   
3         NaN          NaN          NaN          NaN          NaN   
4         NaN          NaN          NaN          NaN          NaN   

   Unnamed: 21  Unnamed: 22  Unnamed: 23  Unnamed: 24  Unnamed: 25  
0          NaN          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN          NaN  
3          NaN          NaN          NaN          NaN          NaN  
4          NaN          NaN          NaN          NaN          NaN  

[5 rows x 26 columns]

In [58]:
col_name = df2.loc[0]
df2 = df2[1:]
df2.rename(columns=col_name,inplace = True)

In [59]:
df2.set_index('customer_id',inplace = True)

In [60]:
df2.head()

first_name  last_name  gender  \
customer_id                                      
1                   Laraine  Medendorp       F   
2                       Eli    Bockman    Male   
3                     Arlin     Dearle    Male   
4                    Talbot        NaN    Male   
5            Sheila-kathryn     Calton  Female   

            past_3_years_bike_related_purchases                  DOB  \
customer_id                                                            
1                                            93  1953-10-12 00:00:00   
2                                            81  1980-12-16 00:00:00   
3                                            61  1954-01-20 00:00:00   
4                                            33  1961-10-03 00:00:00   
5                                            56  1977-05-13 00:00:00   

                          job_title job_industry_category     wealth_segment  \
customer_id                                                                    
1               Executive Secretary                Health      Mass Customer   
2            Administrative Officer    Financial Services      Mass Customer   
3                Recruiting Manager              Property      Mass Customer   
4                               NaN                    IT      Mass Customer   
5                     Senior Editor                   NaN  Affluent Customer   

            deceased_indicator  \
customer_id                      
1                            N   
2                            N   
3                            N   
4                            N   
5                            N   

                                                       default owns_car  ...  \
customer_id                                                              ...   
1                                                           "'      Yes  ...   
2                                 <script>alert('hi')</script>      Yes  ...   
3                                          2018-02-01 00:00:00      Yes  ...   
4            () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No  ...   
5                                                          NIL      Yes  ...   

             NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
customer_id                                                         
1            NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2            NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3            NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4            NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5            NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 25 columns]

In [61]:
df2 = df2.loc[:,df2.columns.notnull()]
df2.head()

first_name  last_name  gender  \
customer_id                                      
1                   Laraine  Medendorp       F   
2                       Eli    Bockman    Male   
3                     Arlin     Dearle    Male   
4                    Talbot        NaN    Male   
5            Sheila-kathryn     Calton  Female   

            past_3_years_bike_related_purchases                  DOB  \
customer_id                                                            
1                                            93  1953-10-12 00:00:00   
2                                            81  1980-12-16 00:00:00   
3                                            61  1954-01-20 00:00:00   
4                                            33  1961-10-03 00:00:00   
5                                            56  1977-05-13 00:00:00   

                          job_title job_industry_category     wealth_segment  \
customer_id                                                                    
1               Executive Secretary                Health      Mass Customer   
2            Administrative Officer    Financial Services      Mass Customer   
3                Recruiting Manager              Property      Mass Customer   
4                               NaN                    IT      Mass Customer   
5                     Senior Editor                   NaN  Affluent Customer   

            deceased_indicator  \
customer_id                      
1                            N   
2                            N   
3                            N   
4                            N   
5                            N   

                                                       default owns_car tenure  
customer_id                                                                     
1                                                           "'      Yes     11  
2                                 <script>alert('hi')</script>      Yes     16  
3                                          2018-02-01 00:00:00      Yes     15  
4            () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No      7  
5                                                          NIL      Yes      8

In [62]:
df2.drop("default",axis = 1,inplace = True)

In [63]:
df2.isna().sum()

first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                  87
dtype: int64

In [64]:
df2.rename({"past_3_years_bike_related_purchases":"other_purchases","job_industry_category":"job_industry"},axis = 1,inplace = True)

In [65]:
df2.head()

first_name  last_name  gender other_purchases  \
customer_id                                                      
1                   Laraine  Medendorp       F              93   
2                       Eli    Bockman    Male              81   
3                     Arlin     Dearle    Male              61   
4                    Talbot        NaN    Male              33   
5            Sheila-kathryn     Calton  Female              56   

                             DOB               job_title        job_industry  \
customer_id                                                                    
1            1953-10-12 00:00:00     Executive Secretary              Health   
2            1980-12-16 00:00:00  Administrative Officer  Financial Services   
3            1954-01-20 00:00:00      Recruiting Manager            Property   
4            1961-10-03 00:00:00                     NaN                  IT   
5            1977-05-13 00:00:00           Senior Editor                 NaN   

                wealth_segment deceased_indicator owns_car tenure  
customer_id                                                        
1                Mass Customer                  N      Yes     11  
2                Mass Customer                  N      Yes     16  
3                Mass Customer                  N      Yes     15  
4                Mass Customer                  N       No      7  
5            Affluent Customer                  N      Yes      8

In [66]:
df2.gender.value_counts()

Female    2037
Male      1872
U           88
M            1
Femal        1
F            1
Name: gender, dtype: int64

In [67]:
df2.gender.replace({"Female":"F","Male":"M","Femal":"F","U":"Not Specified"},inplace=True)

In [68]:
df2.gender.value_counts()

F                2039
M                1873
Not Specified      88
Name: gender, dtype: int64

In [69]:
df2.DOB = pd.to_datetime(df2.DOB).dt.date

In [70]:
pd.to_datetime(df2.DOB).dt.year.mode()

0    1978.0
dtype: float64

In [71]:
df2.DOB.fillna('1978-01-01',inplace = True)

In [72]:
df2.DOB = pd.to_datetime(df2.DOB).dt.date

In [73]:
df2["year"] = df2.DOB.apply(lambda x : str(x).split("-")[0])

In [74]:
df2.year.sort_values()

customer_id
34      1843
720     1931
1092    1935
3410    1940
2413    1943
        ... 
422     2002
2296    2002
2858    2002
3435    2002
1888    2002
Name: year, Length: 4000, dtype: object

In [75]:
#strange error in DOB column with 1843 
df2.drop(34,axis = 0,inplace=True)

In [76]:
df2.columns

Index(['first_name', 'last_name', 'gender', 'other_purchases', 'DOB',
       'job_title', 'job_industry', 'wealth_segment', 'deceased_indicator',
       'owns_car', 'tenure', 'year'],
      dtype='object')

In [77]:
df2.drop(['year'],axis=1)

first_name  last_name         gender other_purchases  \
customer_id                                                             
1                   Laraine  Medendorp              F              93   
2                       Eli    Bockman              M              81   
3                     Arlin     Dearle              M              61   
4                    Talbot        NaN              M              33   
5            Sheila-kathryn     Calton              F              56   
...                     ...        ...            ...             ...   
3996                Rosalia   Halgarth              F               8   
3997                 Blanch     Nisuis              F              87   
3998                 Sarene    Woolley  Not Specified              60   
3999              Patrizius        NaN              M              11   
4000                  Kippy    Oldland              M              76   

                    DOB               job_title        job_industry  \
customer_id                                                           
1            1953-10-12     Executive Secretary              Health   
2            1980-12-16  Administrative Officer  Financial Services   
3            1954-01-20      Recruiting Manager            Property   
4            1961-10-03                     NaN                  IT   
5            1977-05-13           Senior Editor                 NaN   
...                 ...                     ...                 ...   
3996         1975-08-09   VP Product Management              Health   
3997         2001-07-13         Statistician II       Manufacturing   
3998         1978-01-01       Assistant Manager                  IT   
3999         1973-10-24                     NaN       Manufacturing   
4000         1991-11-05    Software Engineer IV                 NaN   

                wealth_segment deceased_indicator owns_car tenure  
customer_id                                                        
1                Mass Customer                  N      Yes     11  
2                Mass Customer                  N      Yes     16  
3                Mass Customer                  N      Yes     15  
4                Mass Customer                  N       No      7  
5            Affluent Customer                  N      Yes      8  
...                        ...                ...      ...    ...  
3996             Mass Customer                  N       No     19  
3997            High Net Worth                  N      Yes      1  
3998            High Net Worth                  N       No    NaN  
3999         Affluent Customer                  N      Yes     10  
4000         Affluent Customer                  N       No     11  

[3999 rows x 11 columns]

In [78]:
df2.last_name.fillna('.',inplace=True)

In [79]:
df2.isna().sum()

first_name              0
last_name               0
gender                  0
other_purchases         0
DOB                     0
job_title             506
job_industry          656
wealth_segment          0
deceased_indicator      0
owns_car                0
tenure                 87
year                    0
dtype: int64

In [80]:
df2.tenure.fillna(np.mean(df2.tenure),inplace = True)

In [81]:
np.mean(df2.tenure)

10.65465235173824

In [82]:
df2.dropna(subset=['job_title','job_industry'],axis = 0,inplace = True)

In [83]:
df2.isna().sum()

first_name            0
last_name             0
gender                0
other_purchases       0
DOB                   0
job_title             0
job_industry          0
wealth_segment        0
deceased_indicator    0
owns_car              0
tenure                0
year                  0
dtype: int64

In [84]:
df2.drop_duplicates(inplace=True)

In [85]:
df2.dtypes

first_name             object
last_name              object
gender                 object
other_purchases        object
DOB                    object
job_title              object
job_industry           object
wealth_segment         object
deceased_indicator     object
owns_car               object
tenure                float64
year                   object
dtype: object

In [86]:
df2 = df2.astype({"other_purchases":int,"tenure":int})

In [87]:
df2.head()

first_name  last_name gender  other_purchases         DOB  \
customer_id                                                             
1              Laraine  Medendorp      F               93  1953-10-12   
2                  Eli    Bockman      M               81  1980-12-16   
3                Arlin     Dearle      M               61  1954-01-20   
9                 Mala       Lind      F               97  1973-03-10   
10            Fiorenze    Birdall      F               49  1988-10-11   

                                        job_title        job_industry  \
customer_id                                                             
1                             Executive Secretary              Health   
2                          Administrative Officer  Financial Services   
3                              Recruiting Manager            Property   
9            Business Systems Development Analyst         Argiculture   
10                        Senior Quality Engineer  Financial Services   

                wealth_segment deceased_indicator owns_car  tenure  year  
customer_id                                                               
1                Mass Customer                  N      Yes      11  1953  
2                Mass Customer                  N      Yes      16  1980  
3                Mass Customer                  N      Yes      15  1954  
9            Affluent Customer                  N      Yes       8  1973  
10               Mass Customer                  N      Yes      20  1988

###  Analyzing the third sheet

In [88]:
df3 = pd.read_excel("../KPMG_VI_New_raw_data_update_final.xlsx",sheet_name='CustomerAddress')
df3

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                           customer_id                                                                                                                              
1                                                     1                                                                                                                              
2                                                     2                                                                                                                              
3                                                     4                                                                                                                              
4                                                     5                                                                                                                              
...                                                 ...                                                                                                                              
3995                                               3999                                                                                                                              
3996                                               4000                                                                                                                              
3997                                               4001                                                                                                                              
3998                                               4002                                                                                                                              
3999                                               4003                                                                                                                              

                     Unnamed: 1 Unnamed: 2       Unnamed: 3 Unnamed: 4  \
0                       address   postcode            state    country   
1            060 Morning Avenue       2016  New South Wales  Australia   
2           6 Meadow Vale Court       2153  New South Wales  Australia   
3            0 Holy Cross Court       4211              QLD  Australia   
4           17979 Del Mar Point       2448  New South Wales  Australia   
...                         ...        ...              ...        ...   
3995            1482 Hauk Trail       3064              VIC  Australia   
3996  57042 Village Green Point       4511              QLD  Australia   
3997     87 Crescent Oaks Alley       2756              NSW  Australia   
3998           8194 Lien Street       4032              QLD  Australia   
3999            320 Acker Drive       2251              NSW  Australia   

              Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9  \
0     property_valuation         NaN         NaN         NaN         NaN   
1                     10         NaN         NaN         NaN         NaN   
2                     10         NaN         NaN         NaN         NaN   
3                      9         NaN         NaN         NaN         NaN   
4                      4         NaN         NaN         NaN         NaN   
...                  ...         ...         ...         ...         ...   
3995                   3         NaN         NaN         NaN         NaN   
3996                   6         NaN         NaN         NaN         NaN   
3997                  10         NaN         NaN         NaN         NaN   
3998                   7         NaN         NaN         NaN         NaN   
3999                   7         NaN         NaN         NaN         NaN   

      Unnamed: 10 

In [89]:
col_name = df3.iloc[0]
df3 = df3[1:]
df3.rename(columns=col_name,inplace = True)


c:\users\punee\appdata\local\programs\python\python39\lib\site-packages\pandas\core\frame.py:4441: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


In [90]:
#removing the nan values
df3 = df3.loc[:,df3.columns.notnull()]

In [91]:
df3.head()

customer_id              address postcode            state    country  \
1           1   060 Morning Avenue     2016  New South Wales  Australia   
2           2  6 Meadow Vale Court     2153  New South Wales  Australia   
3           4   0 Holy Cross Court     4211              QLD  Australia   
4           5  17979 Del Mar Point     2448  New South Wales  Australia   
5           6     9 Oakridge Court     3216              VIC  Australia   

  property_valuation  
1                 10  
2                 10  
3                  9  
4                  4  
5                  9

In [92]:
df3.drop_duplicates()

customer_id                    address postcode            state  \
1              1         060 Morning Avenue     2016  New South Wales   
2              2        6 Meadow Vale Court     2153  New South Wales   
3              4         0 Holy Cross Court     4211              QLD   
4              5        17979 Del Mar Point     2448  New South Wales   
5              6           9 Oakridge Court     3216              VIC   
...          ...                        ...      ...              ...   
3995        3999            1482 Hauk Trail     3064              VIC   
3996        4000  57042 Village Green Point     4511              QLD   
3997        4001     87 Crescent Oaks Alley     2756              NSW   
3998        4002           8194 Lien Street     4032              QLD   
3999        4003            320 Acker Drive     2251              NSW   

        country property_valuation  
1     Australia                 10  
2     Australia                 10  
3     Australia                  9  
4     Australia                  4  
5     Australia                  9  
...         ...                ...  
3995  Australia                  3  
3996  Australia                  6  
3997  Australia                 10  
3998  Australia                  7  
3999  Australia                  7  

[3999 rows x 6 columns]

In [93]:
df3.customer_id.unique() == np.arange(1,4000)

array([ True,  True, False, ..., False, False, False])

In [94]:
#since all values in country column is Australia it doesnt provide any info so we can delete or rename dataframe with australia dataset

In [95]:
df3.state.value_counts()

NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64

In [96]:
df3.drop_duplicates(inplace=True)
#no duplicates to remove

In [97]:
df3.property_valuation.value_counts()

9     647
8     646
10    577
7     493
11    281
6     238
5     225
4     214
12    195
3     186
1     154
2     143
Name: property_valuation, dtype: int64

In [100]:
df.isna().sum()

product_id                 0
customer_id                0
transaction_date           0
online_order               0
order_status               0
brand                      0
product_line               0
product_class              0
product_size               0
list_price                 0
standard_cost              0
product_first_sold_date    0
dtype: int64